compatibility/checker.go

```diff
@@ -0,0 +1,380 @@
+/*
+   Copyright 2020 The Compose Specification Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package compatibility
+
+import "github.com/compose-spec/compose-go/types"
+
+type Checker interface {
```

Comment:
I'm a little worried about manually maintaining all of these.

Would it make more sense to only allow checking a service, network, volume (for example) and let the consumer of the library dive down into the exact specifics?

Reply:
The idea is to offer a full implementation that will reject everything, and compose implementation to "opt-in" for a whitelist / override a few checkers when they need finer control (to check supported values). Doing so, we can introduce new attributes in compose spec, and compose implementations won't have to make changes for those to be ignored (with a warning), just update compose-go. Look at the test case for an illustration sample.

About maintenance, this initial implementation is huge, but subsequent maintenance will just be about introducing a new `Check` method when we introduce a new attribute.

Reply:
That being said I'd like we can generate such code, as well as the compose-go types, from json schema. Do you know any library doing so we could get inspired by?

Reply:
Looking at the tests, consuming this seems to be quite elegant. The only thing that would make it better is consts instead of free strings.

If we have a checklist on new additions to the spect that includes adding the required check code then this is fine.

I don't know of any libraries or tools autogenerating from JSON schema. Maybe @thaJeztah or @silvin-lubecki will though?

Reply:
The JSON schema itself would allow for (e.g.) `"additionalProperties": false` to reject anything that's not defined, but that would mean that each driver would maintain a copy of the schema, and only add the properties they support. Perhaps that would be a two-step process;

1. Use the "upstream" schema to see if the file is well-formed according to the compose spec
2. Use the "implementation" JSON schema to invalidate / warn / error on options that are not supported by the specific implementation


I'm still a bit ... on the huge "interface"; are we using an interface so that implementations are _enforced_ to check for each property?